File: src/Coding_Interview_Patterns_ByteByteGo/ch1_two_pointers/ShiftZerosToTheEnd.java
package Coding_Interview_Patterns_ByteByteGo.ch1_two_pointers;

import static org.junit.jupiter.api.Assertions.*;

/**
 * <a href="https://bytebytego.com/exercises/coding-patterns/two-pointers/shift-zeros-to-the-end">Shift Zeros to the End</a>
 */
public class ShiftZerosToTheEnd {

  public static void main(String[] args) {
    int[] nums = new int[] {0, 1, 0, 3, 2};
    shiftZerosToTheEnd(nums);
    assertArrayEquals(new int[] {1, 3, 2, 0, 0}, nums);

    nums = new int[] {};
    shiftZerosToTheEnd(nums);
    assertArrayEquals(new int[] {}, nums);

    nums = new int[] {0};
    shiftZerosToTheEnd(nums);
    assertArrayEquals(new int[] {0}, nums);

    nums = new int[] {1};
    shiftZerosToTheEnd(nums);
    assertArrayEquals(new int[] {1}, nums);

    nums = new int[] {0, 0, 0};
    shiftZerosToTheEnd(nums);
    assertArrayEquals(new int[] {0, 0, 0}, nums);

    nums = new int[] {1, 3, 2};
    shiftZerosToTheEnd(nums);
    assertArrayEquals(new int[] {1, 3, 2}, nums);

    nums = new int[] {1, 1, 1, 0, 0};
    shiftZerosToTheEnd(nums);
    assertArrayEquals(new int[] {1, 1, 1, 0, 0}, nums);

    nums = new int[] {0, 0, 1, 1, 1};
    shiftZerosToTheEnd(nums);
    assertArrayEquals(new int[] {1, 1, 1, 0, 0}, nums);
  }

  /**
   * Approach: Two Pointers (Unidirectional Traversal) <br>
   * Time Complexity: O(n) <br>
   * Space Complexity: O(1) <br>
   */
  public static void shiftZerosToTheEnd(int[] nums) {
    for (int lastNonZeroIndex = 0, i = 0; i < nums.length; i++) {
      if (nums[i] != 0) {
        int temp = nums[lastNonZeroIndex];
        nums[lastNonZeroIndex] = nums[i];
        nums[i] = temp;
        lastNonZeroIndex++;
      }
    }
  }

//  /**
//   * Approach: Brute Force <br>
//   * Time Complexity: O(n) <br>
//   * Space Complexity: O(n) <br>
//   */
//  public static void shiftZerosToTheEnd(int[] nums) {
//    int[] temp = new int[nums.length];
//    int lastNonZeroIndex = 0;
//    for (int num : nums) {
//      if (num != 0) {
//        temp[lastNonZeroIndex] = num;
//        lastNonZeroIndex++;
//      }
//    }
//    System.arraycopy(temp, 0, nums, 0, nums.length);
//  }
}

File: src/ByteByteGo/Coding_Interview_Patterns/ch01_two_pointers/python3/shift_zeros_to_the_end.py
# from typing import List
#
# '''
# Approach: Two Pointers (Unidirectional Traversal)
# Time Complexity: O(n)
# Space Complexity: O(1)
# '''
#
#
# def shift_zeros_to_the_end(nums: List[int]) -> None:
#   last_non_zero_index = 0
#   for i in range(len(nums)):
#     if nums[i] != 0:
#       temp = nums[last_non_zero_index]
#       nums[last_non_zero_index] = nums[i]
#       nums[i] = temp
#       last_non_zero_index += 1
#
#
# '''
# Approach: Brute Force
# Time Complexity: O(n)
# Space Complexity: O(n)
# '''
#
#
# # def shift_zeros_to_the_end(nums: List[int]) -> None:
# #   temp = [0] * len(nums)
# #   last_non_zero_index = 0
# #   for num in nums:
# #     if num != 0:
# #       temp[last_non_zero_index] = num
# #       last_non_zero_index += 1
# #   for j in range(len(nums)):
# #     nums[j] = temp[j]
